### ISSUE TITLE
Q3 Roadmap 

### ISSUE THREAD

**Issue by dsikka** [vllm-project/speculators#781]
### Status: In Progress

Support multi-node training using the Mooncake backend for hidden states. This initiative also involves updating the project structure to create a dedicated folder / package for various KV Cache connectors (such as the Mooncake connector) and other modifications that could easily be substituted in during training. 

Relevant PRs: 
1. https://github.com/vllm-project/speculators/pull/735
2. https://github.com/vllm-project/speculators/pull/836

### Status: In Progress 

Several improvements have been proposed to extend the vanilla DFlash algorithm, most notably the DSpark algorithm (supported as of: https://github.com/vllm-project/speculators/commit/ff71b1e6c89feb62d77a492075164d207989e538) and the DPace loss function: (supported as of: https://github.com/vllm-project/speculators/pull/736). Q3 will include active work to support these extensions and compare their performances.

Relevant PRs / RFCs:
1. https://github.com/vllm-project/speculators/pull/773
2. https://github.com/vllm-project/speculators/issues/757
3. https://github.com/vllm-project/speculators/pull/736
4. https://github.com/vllm-project/speculators/pull/798
5. https://github.com/vllm-project/speculators/pull/677

### Status: In Progress
1. Long Context Training Support through SP Ulysses: https://github.com/vllm-project/speculators/pull/640
3. Sliding Window Attention for Eagle3 and P-Eagle: https://github.com/vllm-project/speculators/pull/684

### Status: In Progress

Relevant PRs / RFCs: 
1. https://github.com/vllm-project/speculators/issues/652
2. https://github.com/vllm-project/speculators/pull/665

#### Tool Call Improvements
1. https://github.com/vllm-project/speculators/pull/771
2. https://github.com/vllm-project/speculators/pull/769
3. https://github.com/vllm-project/speculators/pull/750
4. https://github.com/vllm-project/speculators/pull/777

### Status: In Progress

With the rapid growth of the project, targeted refactoring and cleanup efforts are needed to simplify and improve areas of the codebase that have become more complex over time. Focusing on improving code organization, removing outdated or redundant implementations, and standardizing existing patterns will improve the overall quality, maintainability, and readability of the codebase.

Areas of improvements include: 
- [ ] Training Args handling (currently at over 70 args)
   - https://github.com/vllm-project/speculators/pull/841
- [ ]  Legacy code removal (Eagle), potentially remove older converters?
- [ ]  Speculators CLI support - currently only exposed through a very large training script but need an API that can be used by RHOAI / KFT within the src code
   -  https://github.com/vllm-project/speculators/pull/1024

### Status: Not Yet Started

This is a general initiative to improve the overall training performance end-to-end using Speculators and vLLM, starting with benchmarking existing performance to identify bottlenecks.

- [ ]  An existing bottleneck is overall support for large hidden state sizes, which can result in timeout issues during training. Potential areas to explore:
   - Reducing sizes in vLLM (projections from the larger hidden sizes should help reduce with the draft models)
   - Quantized hidden states support (e.g. FP8 connector). This will depend on the initiative to have a dedicated  folder / package for various KV Cache connectors, one of which would be specific to FP8.

### DFlash 
1. Mistral 4: https://github.com/vllm-project/speculators/pull/719: https://huggingface.co/RedHatAI/Mistral-Small-4-119B-2603.dflash
 2. Nemotron 3 Ultra: https://huggingface.co/RedHatAI/NVIDIA-Nemotron-3-Ultra-550B-A55B-speculator.dflash
 3. Nemotron 3 Super: https://huggingface.co/RedHatAI/NVIDIA-Nemotron-3-Super-120B-A12B-speculator.dflash

### DSpark 
1. Gemma 4 31B: https://huggingface.co/RedHatAI/gemma-4-31B-it-speculator.dspark
2. GLM 5.2: https://huggingface.co/RedHatAI/GLM-5.2-speculator.dspark
3. Mistral 4: https://huggingface.co/RedHatAI/Mistral-Small-4-119B-2603-speculator.dspark
4. Qwen 3.6 DSpark: https://huggingface.co/RedHatAI/Qwen3.6-35B-A3B-speculator.dspark
